h2o-core/src/main/java/hex/grid/GridSearch.java

```diff
@@ -160,6 +173,125 @@ public long getModelCount() {
+      try {
+        constructScoringInfo(finishedModel);
+        parallelSearchGridLock.lock();
+        grid.putModel(finishedModel._parms.checksum(IGNORED_FIELDS_PARAM_HASH), finishedModel._key);
+        _job.update(1);
+        attemptGridSave(grid);
+
+        grid.update(_job);
+      } finally {
+        parallelSearchGridLock.unlock();
+      }
+      attemptBuildNextModel(parallelModelBuilder, finishedModel);
+    }
+
+    private void onModelBuildFailed(ParallelModelBuilder.ModelBuildFailure modelBuildFailure, ParallelModelBuilder parallelModelBuilder) {
+      parallelSearchGridLock.lock();
+      try {
+        grid.appendFailedModelParameters(null, modelBuildFailure.getParameters(), modelBuildFailure.getThrowable());
+      }finally {
+        parallelSearchGridLock.unlock();
+      }
+      attemptBuildNextModel(parallelModelBuilder, null);
+    }
+
+    private void attemptBuildNextModel(final ParallelModelBuilder parallelModelBuilder, final Model previousModel) {
+      // Attempt to train next model
+      final MP nextModelParams = getNextModelParams(hyperspaceIterator, previousModel, grid);
+      if (nextModelParams != null && isThereEnoughTime() && !_job.stop_requested()) {
+        parallelModelBuilder.run(ModelBuilder.make(nextModelParams));
+      } else {
+        parallelModelBuilder.noMoreModels();
+      }
+    }
+
+    private void constructScoringInfo(final Model model) {
+      ScoringInfo scoringInfo = new ScoringInfo();
+      scoringInfo.time_stamp_ms = System.currentTimeMillis();
+
+      model.fillScoringInfo(scoringInfo);
+      grid.setScoringInfos(ScoringInfo.prependScoringInfo(scoringInfo, grid.getScoringInfos()));
+      ScoringInfo.sort(grid.getScoringInfos(), _hyperSpaceWalker.search_criteria().stopping_metric());
+    }
+
+    private boolean isThereEnoughTime() {
+      final boolean enoughTime = hyperspaceIterator.max_runtime_secs() > 0 && hyperspaceIterator.time_remaining_secs() > 0;
+
+      if (!enoughTime) {
+        Log.info("Grid max_runtime_secs of " + hyperspaceIterator.max_runtime_secs() + " secs has expired; stopping early.");
+      }
+
+      return enoughTime;
+    }
+
+    private MP getNextModelParams(final HyperSpaceWalker.HyperSpaceIterator<MP> hyperSpaceWalker, final Model model, final Grid grid){
+      MP params = null;
+
+      while (params == null) {
+        if (hyperSpaceWalker.hasNext(Optional.ofNullable(model))) {
+          params = hyperSpaceWalker.nextModelParameters(Optional.ofNullable(model));
+          final Key modelKey = grid.getModelKey(params.checksum(IGNORED_FIELDS_PARAM_HASH));
+          if(modelKey != null){
+            params = null;
+          }
+        } else {
+          break;
+        }
+      }
+
+      return params;
+    }
+  }
+
+  /**
+   * Searches the hyperspace and builds models in a parallel way - building the models in parallel.
+   *
+   * @param grid Grid to add models to
+   */
+  private void parallelGridSearch(final Grid<MP> grid) {
```

Comment:
I don't see any support for early stopping anywhere in this implementation, which would make it unusable for `RandomGridSearch`. Did I miss it?

Reply:
Good question - is early stopping thought of @Pscheidl?

Reply:
Early stopping by walker not "vending" other parameters has been added. The same way it works in sequential. It is guaranteed for the walker to receive every model so far + the latest finished. It is then free to stop the vending, because the `stopEarly` method returning true will stop new models from being "queued".

Early stopping by time constraint is tested in `testParallelModelTimeConstraint`. Btw. cancellation by user is also tested.